src/tls/crl.rs

```diff
@@ -0,0 +1,466 @@
+        inner.crl_data = der_crls;
+        inner.revoked_serials = new_revoked_serials;
+        inner.last_load_time = Some(SystemTime::now());
+
+        debug!(
+            "CRL loaded successfully ({} CRL(s), {} total revoked certificate(s))",
+            inner.crl_data.len(),
+            total_revoked
+        );
+        Ok(has_new_revocations)
+    }
+
+    /// Parse PEM-encoded CRL data that may contain multiple CRL blocks
+    /// Returns a Vec of DER-encoded CRLs
+    fn parse_pem_crls(pem_data: &[u8]) -> Result<Vec<Vec<u8>>, CrlError> {
+        let data_str = std::str::from_utf8(pem_data)
+            .map_err(|e| CrlError::ParseError(format!("Invalid UTF-8: {}", e)))?;
+
+        let mut crls = Vec::new();
+        let mut in_pem = false;
+        let mut base64_data = String::new();
+
+        for line in data_str.lines() {
+            if line.starts_with("-----BEGIN") {
+                in_pem = true;
+                base64_data.clear(); // Start new CRL block
+                continue;
+            }
+            if line.starts_with("-----END") {
+                if in_pem && !base64_data.is_empty() {
+                    use base64::Engine;
+                    let der = base64::engine::general_purpose::STANDARD
+                        .decode(&base64_data)
+                        .map_err(|e| {
+                            CrlError::ParseError(format!("failed to decode base64: {}", e))
+                        })?;
+                    crls.push(der);
+                    base64_data.clear();
+                }
+                in_pem = false;
+                continue;
+            }
+            if in_pem {
+                base64_data.push_str(line.trim());
+            }
+        }
+
+        if crls.is_empty() {
+            return Err(CrlError::ParseError(
+                "no valid CRL blocks found in PEM data".to_string(),
+            ));
+        }
+
+        Ok(crls)
+    }
+
+    /// Verify CRL validity period
+    fn verify_crl_validity(
+        crl: &x509_parser::revocation_list::CertificateRevocationList,
+        allow_expired: bool,
+    ) -> Result<(), CrlError> {
+        let now = SystemTime::now();
+        let unix_now = now
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs() as i64;
+
+        // check thisUpdate (CRL issue time)
+        if unix_now < crl.tbs_cert_list.this_update.timestamp() {
+            warn!("CRL is not yet valid");
+        }
+
+        // check nextUpdate (CRL expiry)
+        if let Some(next_update) = &crl.tbs_cert_list.next_update
+            && unix_now > next_update.timestamp()
+        {
+            if !allow_expired {
+                return Err(CrlError::ExpiredCrl);
+            }
+            warn!("CRL is expired but allow_expired_crl is enabled");
+        }
+
+        Ok(())
+    }
+
+    /// Check if any certificate in the chain is revoked
+    pub fn is_revoked_chain(
```

Comment:
Is there a reason we didn't use something like https://docs.rs/rustls-webpki/latest/webpki/enum.CertRevocationList.html#method.find_serial?

Reply:
Ah. This is better. validating CRL with rustls-webpki. Thanks for the pointer.